### ISSUE TITLE
Emoji data for Scribe apps

### ISSUE THREAD

**Issue by andrewtavis** [scribe-org/Scribe-Data#14]
### Terms

- [X] I have searched [open and closed feature requests](https://github.com/scribe-org/Scribe-Data/issues?q=is%3Aissue+label%3Afeature)
- [X] I agree to follow Scribe-Data's [Code of Conduct](https://github.com/scribe-org/Scribe-Data/blob/main/.github/CODE_OF_CONDUCT.md)

### Description

Now that [Scribe-iOS](https://github.com/scribe-org/Scribe-iOS) is adding autosuggest in [#194](https://github.com/scribe-org/Scribe-iOS/issues/194) and autocomplete in [#188](https://github.com/scribe-org/Scribe-iOS/issues/188), other additions for these features could be considered including emojis. This feature would be to create python scripts that would create arrays of words and emojis that can represent them. When one of these words is entered, the emojis could then serve as autosuggestions or completions (depending on if the user has pressed space or not respectively).

These scripts would make use of [emoji](https://github.com/carpedm20/emoji/) for the representations of emojis as words in the given languages, and would also make use of [nltk](https://github.com/nltk/nltk) for stemming (not lemmatizing) words to derive `smile` from `smiling`. Other language packages could be leveraged to derive adjective forms like `smiley`, and convert words like types of trees (aspen, birch, etc) to "tree" so they can also be converted (likely another issue).

### Contribution

I'd be happy to work with someone on this once the features in Scribe-iOS are finished :)

The corresponding Scribe-iOS issue for this is [#51](https://github.com/scribe-org/Scribe-iOS/issues/51), which would need to be worked on in the foreseeable future for this issue to make sense.

**Comment by andrewtavis** [comment]
Note that final version of this would be included in [update_data.py](https://github.com/scribe-org/Scribe-Data/blob/main/src/scribe_data/load/update_data.py) so that changes to the scripts would be reflected in the apps.

**Comment by andrewtavis** [comment]
Adding some easter eggs into this feature could also be a nice touch. These might be best as just autosuggestions and not corrections. Some ideas are:

- Wikidata: 💚 and 📁
- Wikipedia: ❤️ and 📚
- WIkimedia: 🌐 and 📖
- Scribe: 💙 and ✏️

**Comment by wkyoshida** [comment]
Hey @andrewtavis - I'd love to help out with this!

_I presume that this is unblocked now with the completion of https://github.com/scribe-org/Scribe-iOS/issues/194 and https://github.com/scribe-org/Scribe-iOS/issues/188 - would that be correct?_

Looking into it, I had some initial thoughts (mostly regarding the source for emoji data):
- Maybe I overlooked something, but would [emoji](https://github.com/carpedm20/emoji/) be able to really only provide Scribe the most (or two most) common "short name"[1] perhaps? 
- In its README, [emoji](https://github.com/carpedm20/emoji/) links to [EmojiTerra](https://emojiterra.com/), which could be an interesting resource in that it has both "short names" and "keywords"[2] for emojis in multiple languages. I'm thinking that "short names" and "keywords" could be leveraged as the "trigger words" [3].
  - :heavy_check_mark: With data for multiple languages, EmojiTerra could be useful for implementation across languages.
  - :x: However, after some browsing, my feeling was that it does lack perhaps in the quantity of "keywords" per emoji. For instance, :running: could perhaps also be associated with the word "exercise". 
- I did come across [emojilib](https://github.com/muan/emojilib) though, which has an interesting [emoji-en-US.json](https://github.com/muan/emojilib/blob/main/dist/emoji-en-US.json) with a larger quantity of "keywords".
  - :heavy_check_mark: The larger amount of associated "keywords" could translate into more "trigger words" that could be used.
  - :heavy_check_mark: The project looks like it has a neat [contribution process](https://github.com/muan/emojilib/blob/main/CONTRIBUTING.md) to add/edit/delete "keywords".
  - :x: The project appears to only have US English presently. Perhaps the Scribe translation data could be used to get the equivalent in other languages, but literal translations are likely not ideal across-the-board.

With all that said, I'd love to continue discussions on this. Curious on further thoughts!

P.S. I do like the idea of using [nltk](https://github.com/nltk/nltk) for working with word stems :+1: 

---
Clarification on some terminology above:
- [1] **short name**: (or shortcode) a shorthand qualifier used to insert an emoji in supported applications, e.g. in GitHub, the short name `:+1:` renders to :+1:
- [2] **keywords**: words that have association with the emoji, e.g. the `:running:` emoji (:running:) also has the keyword "marathon" associated with it in [EmojiTerra](https://emojiterra.com/person-running/)
- [3] **trigger words**: words that would prompt an emoji as an option for the auto-suggestion or auto-completion Scribe functionality

**Comment by andrewtavis** [comment]
> Hey @andrewtavis - I'd love to help out with this!

Hey @wkyoshida, would be great to get your help with this, and as always just with your research you already have! 😊 Thanks for your efforts already - especially with all the formatting that you do that makes it quite easy to follow 🙏

> I presume that this is unblocked now with the completion of [scribe-org/Scribe-iOS#194](https://github.com/scribe-org/Scribe-iOS/issues/194) and [scribe-org/Scribe-iOS#188](https://github.com/scribe-org/Scribe-iOS/issues/188) - would that be correct?

This is now unblocked, correct 😊

> * Maybe I overlooked something, but would [emoji](https://github.com/carpedm20/emoji/) be able to really only provide Scribe the most (or two most) common "short name"[1] perhaps?

Would be something we'd need to investigate, but there could be a way to do this 👍

> * ❌ However, after some browsing, my feeling was that it does lack perhaps in the quantity of "keywords" per emoji. For instance, 🏃 could perhaps also be associated with the word "exercise".

I agree with the positive point, and would say that for now let's not worry too much about the edge cases in the planning phase :) There could be a way to only use keywords for very popular emojis or subset in a logical way as we don't need autosuggestions for all emojis necessarily. With Scribe the big thing is keep single edge cases that we're dealing with to a minimum. There might even be some metadata on emojis on Wikidata that we could use 💡🤔 Big thing is that keywoards are going to be very necessary for some, as your example of `:+1:` for 👍 indicates, as that would really only be useful if it's linked to `good` and the language equivalents.

> * ❌ The project appears to only have US English presently. Perhaps the Scribe translation data could be used to get the equivalent in other languages, but literal translations are likely not ideal across-the-board.

I'd say let's limit ourselves to [emoji](https://github.com/carpedm20/emoji/) and NLTK for now so that we're not overcomplicating it all :) :) This would definitely be something to look into if multilingual support grows or when Scribe adds the sought after [English keyboard](https://github.com/scribe-org/Scribe-iOS/issues/7). I'd also sadly suggest that we not use the translations here for anything too serious. The whole point of those now is to get the functionality in, as it's very much `beta`. I'm in talks with folks/the community at Wikidata to try to get translations in, and when that happens we'll be able to really make use of them. As of now they're machine translations from [🤗 Hugging Face](https://github.com/huggingface), and with that have their problems :)

Where to from here? 😊 I'm away for the weekend, just FYI, but if you wanted to start exploring the generation of an `emojis.json` file you'd be welcome. I'd also be happy to setup a base file as well, but you'd also be welcome to make `gen_emoji_triggers.py` to [src/scribe_data/load](https://github.com/scribe-org/Scribe-Data/tree/main/src/scribe_data/load). To me it makes sense that this functions like autosuggest where we'd work on functions that are loaded into Jupyter right now, as seeing the outputs is going to be very important :)

**Comment by wkyoshida** [comment]
Hey @andrewtavis! I meant to reply earlier, but wanted to focus on the keyboard variants discussion first as that was higher priority. I can definitely help with the development for this though! 

Resuming: 

> Would be something we'd need to investigate, but there could be a way to do this :+1:

For some added clarification, my initial point, perhaps, was more so cause the [data file](https://raw.githubusercontent.com/carpedm20/emoji/master/emoji/unicode_codes/data_dict.py) that the [emoji](https://github.com/carpedm20/emoji/) project makes use of appears to be limited to only 1 "short name" per human language (some emojis have 2 for English). Browsing the file, to me it felt a little limiting if just going off of [emoji](https://github.com/carpedm20/emoji/), since, as a user, some words that I'd expect would trigger an emoji, would not. On the other hand though, if we look to really just get an MVP out with this issue, then I think that it'd probably be fine. Trigger word improvements could happen over time.

>There might even be some metadata on emojis on Wikidata that we could use :bulb::thinking:

:+1: One reason why [EmojiTerra](https://emojiterra.com/) piqued my interest was the good amount of data that it has, including for multiple languages. It also has more "keywords", which I thought could make it a little less limiting than only going off of [emoji](https://github.com/carpedm20/emoji/). With that said, one thought that I had was that Scribe could try reaching out to the EmojiTerra team even ([their contact](https://emojiterra.com/contact/)) and asking if they'd be interested in helping populate WikiData. No idea if they would, but just a thought.

> I'd also sadly suggest that we not use the translations here for anything too serious.

Completely agree. My earlier note was added more as a potential path, but I too think that the Scribe translation data shouldn't be used for this. As you pointed out, it is still very much in `beta`.

> would be happy to set up some base codes for this in the coming days :)

Sounds great! That'd be welcomed. And again, I can definitely take it from there and help with development :+1: :rocket:

**Comment by andrewtavis** [comment]
Great to talk to you about all this, @wkyoshida 😊

> On the other hand though, if we look to really just get an MVP out with this issue, then I think that it'd probably be fine. Trigger word improvements could happen over time.

Let's start down this path, and maybe we'll discover that there are ways of expanding it that are easier :) This tends to be how things work around here, as we planned to just do autocomplete, and then it snowballed into that with autosuggest and everything else that v2.0 had ☃️😅


> With that said, one thought that I had was that Scribe could try reaching out to the EmojiTerra team even ([their contact](https://emojiterra.com/contact/)) and asking if they'd be interested in helping populate WikiData. No idea if they would, but just a thought.

The detail you put into all this is much appreciated 🤝 Looking into EmojiTerra makes sense :) Let's definitely consider this as the main source for the extra meanings, and reaching out to them about a collaboration or working with Wikidata would also be great! Is there an effective way that we can access the data on EmojiTerra, or would the hope be that a migration to Wikidata happens? I can reach out to the Wikidata community about emojis too as they might be willing to help with a WIP query :)

> Sounds great! That'd be welcomed.

I'll work on the base file for you this weekend! Hope your week is going well so far 😊

**Comment by wkyoshida** [comment]
> This tends to be how things work around here, as we planned to just do autocomplete, and then it snowballed into that with autosuggest and everything else that v2.0 had :snowman_with_snow::sweat_smile:

:laughing: Yeah - alright, I think that sounds fair. Let's go with an MVP, and improvements can then happen over time.

> Is there an effective way that we can access the data on EmojiTerra, or would the hope be that a migration to Wikidata happens?

That would be the gap, I think. I'm not sure if I saw an easier way to access the data or either the source code for EmojiTerra. I was hoping a dump to Wikidata could be done, so Scribe could maintain its "powered-by-Wikimedia" status throughout the data it uses. As a last resort, of course, Scribe could always just scrape EmojiTerra, though I don't think that would be the ideal path.

> I can reach out to the Wikidata community about emojis too as they might be willing to help with a WIP query :)

That would be awesome! :raised_hands: 

> Hope your week is going well so far :blush:

Oh for sure - thanks! Hope yours is as well :v:

**Comment by wkyoshida** [comment]
Hey @andrewtavis, given the discussions had in https://github.com/scribe-org/Scribe-iOS/issues/89 (mainly referring to the idea of using a DB on the back-end), should development for this issue hold off a bit? If the data structure/storage changes, I'm thinking whatever scripts get created would have to be modified anyways (which could likely be true for all other data extraction scripts). As this issue isn't high priority though, I was thinking it could perhaps wait (at least until it's determined if the data structure will change). Curious what your thoughts are on this as well.

However, I am also thinking that Scribe could still follow through with trying to contact the EmojiTerra team - to at least get a little ahead. It could happen that by the time Scribe is ready to go develop on this issue, the data is already in Wikidata! :raised_hands: One can dream! :laughing: If you'd like to delegate it off, I could give it a shot to contact EmojiTerra. It could also make sense to have yourself do it though as the main Scribe dev and representative. Just let me know!

**Comment by andrewtavis** [comment]
I think it does make sense to hold off on this for now, but reaching out to them could still be beneficial in the short term :) Do you want to reach out to me in my email that's in my profile about that? We can draft an email and I can CC you in it? I think it would make sense that it comes from me, but I'd be happy to loop you in 😊

Btw: new feedback from one of my cowers who tried Scribe today is that he'd like QWERTY for all keyboards, including German. Looks like it'd make sense to have that option available for all keyboards :) :)

**Comment by andrewtavis** [comment]
@wkyoshida, https://github.com/scribe-org/Scribe-iOS/issues/241 is the issue opened by a coworker about the keyboard layouts :) Well done predicting the importance of this 👏😊

**Comment by andrewtavis** [comment]
> Sounds great, @andrewtavis! I sent you an email just now. Let's draft something up 👍

Will write back later in the week, @wkyoshida 😊🚀

> I still think Scribe ~could~ **should** provide the customization options, but the need or demand for it might not be the same as in Scribe-iOS or Scribe-Android. Little sidetrack, but just a thought I had.

Glad that you're keeping Scribe-Desktop in mind! Yes it'll definitely be more about keyboard shortcuts, so this won't be as much of an issue at first, but then for that the speed of it all will be important as it's easy enough to have Google Translate open or go to a website, so we need to give those instantaneous results and keep people in their workflow 💪

> Edit: To clarify, I think the ability to customize Scribe-Desktop keyboards to all use the same layout would be beneficial though

You bring up a good point here. Some people are using their keyboards with keys that are not exactly mapped to their keyboards - specifically our users working in a foreign country 😊 Wouldn't we be reading in the characters that the user is typing rather than the keys that they're pressing though? Not sure 🤔 Your thoughts on this would be very welcome :)

**Comment by wkyoshida** [comment]
> Will write back later in the week

Sounds good!

---
I also replied to the points made here about Scribe-Desktop over on an issue on that repo instead, [this discussion](https://github.com/scribe-org/Scribe-Desktop/issues/1#issuecomment-1291492398). Wanted to make sure we didn't derail too much from the discussion here about the emoji data :sweat_smile:

**Comment by andrewtavis** [comment]
Checking in with you here as I said via email, @wkyoshida, as it makes sense to keep it in GitHub if we won't reach out to EmojiTerra. Looking into the Unicode and EmojiTerra's sources a bit more, I came across [Unicode CLDR](https://github.com/unicode-org/cldr), and specifically [their annotations](https://github.com/unicode-org/cldr/tree/a925b7d8894781ca1cdda798fcb5ac575fa9e4be/common/annotations) and [cldr-json files](https://github.com/unicode-org/cldr-json).

That looks to be everything we'd want, but then we'll need to check licensing and figure out if there's a good endpoint for all this 😊 Makes sense that there's someplace in code that all this lives :) :)

**Comment by wkyoshida** [comment]
Ah-ha! Alright - great find, @andrewtavis :raised_hands: 

Eventually I was able to get there also :laughing: but I found Unicode's stuff perhaps not as easy to navigate/browse through imo (not sure if you felt the same). Had to move around subdomains under `unicode.org` to find things.

Anyways - I think those make sense! There does appear to be the [ICU](https://github.com/unicode-org/icu), which has official C/C++ and Java libraries that could be used to work with the CLDR data. There is also a list of related [wrapper projects](https://icu.unicode.org/related) for other languages that Unicode links to as well. We could look into those.

**Comment by andrewtavis** [comment]
Ya the Unicode stuff was very hard to get through, @wkyoshida 😅 I was at one point trying to put `/de/` into parts of the web domain in order to try and find something German 🤣

The ICU looks like a good path going forward :) We can maybe look first into [pyicu](https://gitlab.pyicu.org/main/pyicu) to see if it can work. I see from your profile that you have C/C++ and Java experience, but maybe trying to keep it all under one language right now would be best?

**Comment by wkyoshida** [comment]
> I was at one point trying to put `/de/` into parts of the web domain in order to try and find something German :rofl:

:laughing: 

> maybe trying to keep it all under one language right now would be best?

Oh yes! I would agree. I didn't mean to sound like I was leaning for the C/C++ or Java ones, if it did :laughing: Only mentioned them, since they appear to be the officially supported ones by Unicode. I was thinking that the Python wrapper would make sense as well. It does look like it is actively maintained, and according to this [statement](https://gitlab.pyicu.org/main/pyicu#api-documentation), I would hope pretty on-par with whatever is available in the official libraries? Keeping with Python would make sense as that's what the other scripts in Scribe-Data already use.

**Comment by andrewtavis** [comment]
So we have ourselves a decision on this! 😊 Fantastic. Let's read through the docs, and I'll make us a notebook to work from tomorrow :)

**Comment by wkyoshida** [comment]
Hey @andrewtavis, real quick just to clarify something for me.

Also sharing some context from our quick email chat just so it's here for others to understand: 
> I reached out to a contact within the Wikidata community and he agreed that the Wikidata support for emojis in other languages is pretty weak, but then he suggested we go directly from Unicode in this case as they have the endpoints for people to use.

From the above, was it suggested that Scribe could use Unicode directly for the emoji data or that Scribe could use Unicode to fill in the gaps in Wikidata (so Scribe is then able to later use the data from Wikidata also)? Mostly asking, because I'm thinking that, with the former, implementing something now could still run into that earlier point that we discussed of potentially having to modify the data extraction scripts later. That would be due to a possible change in the data solution. The question then, I guess, could be: Is the emoji functionality a priority enough today that Scribe is okay with perhaps dealing with data extraction rework later? Or is there enough reason to first more concretely discuss the long-term strategy for the data solution's back-end? We can have discussions in https://github.com/scribe-org/Scribe-iOS/issues/89 as well, wherever the topic makes the most sense. For transparency, I don't mean to sound definitively opposed to working on emoji data _atm_ :laughing: more so just trying to better understand in terms of priority and roadmap.

**Comment by andrewtavis** [comment]
The suggestion was that we just go directly from Unicode. This would be a mass upload, and this is something that the Wikidata community is very much against. We'd need to be willing to administer the data that was added at the very least, but even then it likely wouldn't get their blessing as the content to admin ratio would be skewed by this.

I'd say we still do something now, and we can switch it over to Wikidata later when there's more support. Or maybe we just won't. There's no necessity that we have to use Wikimedia for everything :) With there being such a universal solution for working with Unicode data, it could be years until it's properly implemented in Wikidata.

Thanks for checking on this! Will get to that script for us to explore a bit in the next few days 😊

**Comment by wkyoshida** [comment]
Sounds good!

Gotchu - that makes sense to me :+1: 
One other point that I was also more wondering in regards to: 
> Or is there enough reason to first more concretely discuss the long-term strategy for the data solution's back-end?

is on the Scribe data-hosting side - related to the [discussion](https://github.com/scribe-org/Scribe-iOS/issues/89#issuecomment-1288076716) we were having on using GitHub vs Toolforge vs something else, etc.. I was more referring, I guess, to the idea of potentially leveraging a DB on the back-end side also as opposed to JSONs in the Scribe-Data repo as the centralized location where Scribe data is hosted. Are you thinking as well of already going forward with getting the emoji data before? I'm thinking that would be fine; just wanted to throw the following for consideration as well. 

For any future data extraction work - be it emoji data, new keyboard data, translation data, etc. - could it make sense to prioritize figuring out the centralized data-hosting first, since moving to a DB could impact how Scribe handles, stores, and provisions data? 

How I think I am understanding, prioritized Scribe work is: 
1. main priorities today: 
    - iOS SQLite
    - app menu
2. data download via GitHub

After that, however, I see:
- the rest of the [`-priority-`](https://github.com/scribe-org/Scribe-iOS/issues?q=is%3Aopen+is%3Aissue+label%3A-priority-) work, includes new keyboards and their needed data
- look into possible data download via a non-GitHub option (perhaps Toolforge)
  - I'm more wondering, I guess, where this investigation/brainstorm fits into the roadmap, since how Scribe data is handled could affect how new keyboard data for the above point and also emoji data for this issue are handled

I might also have some of the work prioritization misunderstood; so please keep me honest! :laughing: If getting in the emoji feature and/or new keyboards is priority over investigating a possible non-GitHub data-hosting option, let me know.

**Comment by andrewtavis** [comment]
> Are you thinking as well of already going forward with getting the emoji data before?

I'm generally in favor of it :) Doesn't have to happen, but it makes sense to me that we get scripts that give us the JSON versions within SQLite, we understand what we need to make available via downloads, and then we work towards that. Downloads will be tested as patches between versions before the full system is integrated, so it makes sense that it would follow this order 😊

> How I think I am understanding, prioritized Scribe work is

Your points mostly make sense to me for the next ordering of the major functionality. I think we're fine with the keyboards we have now until we start gaining traction with users and institutions pushing Scribe as a viable option for second language learners. For that we need to fix the bugs (data issue - v2.2), add some customization (menu - v2.3ish), and finish any missing features (emojis and some others from the priority issues like maybe the user onboarding tutorial). Then from there we have data downloads (v3.0), and from there we switch the autosuggest over to machine learning algorithms and add in autocorrect (v4.0) 🚀

Let me know what you think on the above! :)

**Comment by andrewtavis** [comment]
@wkyoshida, 4a9a392 adds [gen_emoji_words.ipynb](https://github.com/scribe-org/Scribe-Data/blob/main/src/scribe_data/load/gen_emoji_words.ipynb) that we can make the commits to this to. We can work in there until we have some finalized functions, and then those can be migrated into Python files and loaded into the notebook (hence the `sys.path.append`) :) Wanted to send it along in case you wanted to start. Will look into [pyicu](https://gitlab.pyicu.org/main/pyicu) in the coming days 😊

**Comment by wkyoshida** [comment]
> ... it makes sense to me that we get scripts that give us the JSON versions within SQLite... Downloads will be tested as patches between versions before the full system is integrated

Ah yeah okay - I'm seeing more what you likely have in mind. I think what was happening is I was considering the possibility of a Scribe architecture not dependent on JSON for its data at all - only the DBs. But you are right in that there could definitely be a phase where Scribe is leveraging both JSON and SQLite in different places, perhaps likely integrated via SQLite's [JSON1](https://www.sqlite.org/json1.html). Non-dependence on JSON could eventually happen, but it also does not have to nor does it have to happen soon.

> ... I think we're fine with the keyboards we have now... Let me know what you think on the above! :)

The one keyboard I am thinking though that could be good for Scribe to revisit is the English one, https://github.com/scribe-org/Scribe-iOS/issues/7. With English being such a popular second language and dominant in global communication, I feel that the need is definitely there (also evidenced by the user vote in the issue already). Scribe would need to revisit where it could fit into the roadmap, but I'm leaning on the opinion that it could even come after the customization (menu - v2.3ish), since that is the feature mostly blocking it. The other features - I'm more seeing with the perspective that they're _nice-to-haves_, but they don't necessarily impede a user from being able to already benefit from Scribe. I feel that the English keyboard though is definitely a core feature that is missing. Curious what your thoughts are. It appears that the idea for English has been out for a while; so could very well be that considerations for it have changed :smile: 

> @wkyoshida, https://github.com/scribe-org/Scribe-Data/commit/4a9a392bceb85578c4441846a48916f3a40ca5fa adds [gen_emoji_words.ipynb](https://github.com/scribe-org/Scribe-Data/blob/main/src/scribe_data/load/gen_emoji_words.ipynb)... Wanted to send it along in case you wanted to start.

Thanks! Sounds great, @andrewtavis  :+1: I'll give it a look.

**Comment by andrewtavis** [comment]
> I think what was happening is I was considering the possibility of a Scribe architecture not dependent on JSON for its data at all - only the DBs.

This would definitely be a long term goal, but I think it makes sense that the interim is that we're sending a file with the data to be updated. Happy to be proved wrong and jump right into direct SQLite updates though 😊


> The one keyboard I am thinking though that could be good for Scribe to revisit is the English one, [scribe-org/Scribe-iOS#7](https://github.com/scribe-org/Scribe-iOS/issues/7). With English being such a popular second language and dominant in global communication, I feel that the need is definitely there (also evidenced by the user vote in the issue already).

Will write back on this in the other issue, but I definitely agree that it should be a focus :)

**Comment by andrewtavis** [comment]
Hey @wkyoshida 😊 I'm looking into this a bit more now that #21 is all done. Basic thing I can say is that [pyicu](https://gitlab.pyicu.org/main/pyicu) doesn't seem to be what we want. We're looking for the CLDR data specifically, and that package doesn't have access by the looks of it 🤔

Looking at it, if we're looking for the fastest/easiest way to do this, it might be to simply download the files we need from [unicode-org/cldr-json](https://github.com/unicode-org/cldr-json) and write the appropriate Python script to invert the JSONs into word -> emoji pairs. As of now it doesn't look like there's a good Python tool, but then I might be missing it. 

What do you think of this quicker solution? :)

**Comment by wkyoshida** [comment]
> This would definitely be a long term goal, but I think it makes sense that the interim is that we're sending a file with the data to be updated. Happy to be proved wrong and jump right into direct SQLite updates though :blush:

I'm thinking that a scenario that could call for Scribe to jump more aggressively into SQLite is if downloads actually get too large. As we've discussed, an idea could be that only the _diff_ gets downloaded and Scribe leverages something like a `last_updated` field on the data. However, I guess a `last_updated` field could also just be added to the JSON files as well; so this might not be justification enough to jump right in. The trick, I'm thinking, will be more dependent actually on how the data extraction could keep track of what data is new/changed and on the existence of a (possibly-Toolforge) data hosting back-end to do the calculation of these data _diffs_.

> Basic thing I can say is that [pyicu](https://gitlab.pyicu.org/main/pyicu) doesn't seem to be what we want. We're looking for the CLDR data specifically, and that package doesn't have access by the looks of it :thinking:

Yeah.. I took a look as well and wasn't really finding what Scribe needs either, which is unfortunate. I'm also thinking that perhaps just downloading the [unicode-org/cldr-json](https://github.com/unicode-org/cldr-json) files might be the path-forward here. I have a potential idea on how to do it; I'll open up a PR shortly for it.

**Comment by andrewtavis** [comment]
> I'm thinking that a scenario that could call for Scribe to jump more aggressively into SQLite is if downloads actually get too large. As we've discussed, an idea could be that only the _diff_ gets downloaded and Scribe leverages something like a `last_updated` field on the data.

Would be 100% fine with this :) Big thing is that I think it makes sense that we never have a direct connection to Wikidata, but instead are keeping the prepared data somewhere and accessing it. Specifically what I mean is that we're giving people pre-prepared packs, not that one person has the most up to date version of the data whenever they do the update. This is how it works for apps like Open Street Maps wrappers, so I think we should follow this example 😊

> I have a potential idea on how to do it; I'll open up a PR shortly for it.

Great! Looking forward! 🚀🚀 Anything code based that allows us to update the JSONs we have locally within a workflow would be very very welcome 😊

**Comment by andrewtavis** [comment]
Also, @wkyoshida, [3c111c9](https://github.com/scribe-org/Scribe-Data/commit/3c111c92d0eb4ec1f8b51450a74c0b22ea543783) added a roadmap to the readme (I also added it to iOS). I think it makes sense to include it :) Let me know if you have any feedback on it 😊

**Comment by wkyoshida** [comment]
> [3c111c9](https://github.com/scribe-org/Scribe-Data/commit/3c111c92d0eb4ec1f8b51450a74c0b22ea543783) added a roadmap to the readme (I also added it to iOS). I think it makes sense to include it :)

I agree! :raised_hands: 
I'm thinking it'll undoubtedly be helpful for contributors and users to understand the direction the project is taking :+1:
Just a quick thought, but do you have any thoughts on using the [Projects](https://github.com/orgs/scribe-org/projects?query=is%3Aopen) tab of the Scribe organization? Can't say I've used it a ton before, but I was thinking it could be organized to hold the aggregate roadmap with issues and PRs from **all repos** together - which could make sense since work on different repos might happen concurrently. The new roadmap section in the READMEs could also potentially just link to the project (might help with not having to update all READMEs whenever something changes). 

> Main question I have as far as version IDs are concerned is whether adding the ability to choose the base language the user translates from is a major release? As of now I have it as v2.4.0, but then we'll doubtless be doing lots of other features once the menu's done, and even that by itself is maybe 3.0 due to all the refactoring :thinking:

That's a good thought - hmm. I'm thinking that it potentially could as well actually. 

It can also get tricky thinking what version number features will end up falling under _imo_ :laughing: If the organization `Projects` tab ends up getting leveraged, would you have any thoughts on organizing/assigning the issues with some sort of priority value? I'm thinking that it could still show a general roadmap for all things Scribe, while also not boxing-in features into version numbers prematurely. Versioning can get assigned as features completed, I think, especially since misc. features can also get completed here and there.

As far as what is on the roadmap though - I think it makes sense :+1:

**Comment by andrewtavis** [comment]
> Just a quick thought, but do you have any thoughts on using the [Projects](https://github.com/orgs/scribe-org/projects?query=is%3Aopen) tab of the Scribe organization?

We have [projects for iOS](https://github.com/scribe-org/Scribe-iOS/projects?type=classic), but as of now I haven't really found them useful... I think that as the team grows the organization wide ones would definitely be the way to go though 🚀 I'm also happy to implement them now if you'd find them helpful, and agree that referencing a project for a point on the roadmap would be more descriptive :)

> That's a good thought - hmm. I'm thinking that it potentially could as well actually.

I'll update it to v3.0.0 right now, but let's definitely keep in mind that this is a WIP presentation of what we want to work on 😊 Happy to expand on it later as we implement projects and other organization tools :)

**Comment by wkyoshida** [comment]
Sounds good, @andrewtavis :+1: :raised_hands:  

Regarding the `Projects` tab, I did notice before too there was some in the iOS repo. I guess I was more thinking of revisiting the idea of using `Projects` because:
- As already mentioned, the organization-level one could represent the aggregate roadmap more easily
- There is new functionality. The projects in Scribe-iOS appear to be using the `Classic` features; so I wasn't sure if you've already explored the functionality with the new `Projects` and saw any potential there. 
  - It was mainly this actually, I think. There seems to be **a lot** more features with the new version that make it easier to organize issues. 

I don't think there's a dire need for it - only thought of this because referencing an issues board, in my mind, seemed easier to reference and work with than a README section.

**Comment by andrewtavis** [comment]
This does make sense, @wkyoshida 😊 I haven't really looked into the features of the new projects, and referencing that would be easier than constantly needing to update the readme based on how things change :) At least writing it we have a better idea of how to structure the projects going forward 🤓 

Will remove the projects from iOS as a first step, and then we can go from there. You'd suggest that all the projects be at the organization level, right? Makes sense to me that we have one place for organizing the various elements, as there doubtless are ways to filter the projects to see what you're interested in :)

**Comment by wkyoshida** [comment]
> You'd suggest that all the projects be at the organization level, right?

Yea :v: Also, since I'm thinking that related work across repos can be under the same groupings too - one can easily see they go together. Like with the below where there is both Scribe-iOS and Scribe-Data work:

> - Scribe-iOS v3.0.0 and Scribe-Data v3.0.0
>   - Expand translation data and add English keyboard
>   - Main issues: Scribe-iOS #&#8203;7 and Scribe-Data #&#8203;23

**Comment by andrewtavis** [comment]
Sounds great 😊 Give me a few days and I'll try to get them up and running 🚀

**Comment by andrewtavis** [comment]
Brining the discussion back in here, @wkyoshida :) What should we look to do now? It looks like we might be able to work on the basis of https://github.com/scribe-org/Scribe-iOS/issues/51 right now, as the main focus here is getting the order down given popularity? 

Plan is to definitely include all of this in v2.2.0, btw 😊 Really looking forward to adding such a key feature!

**Comment by andrewtavis** [comment]
Looking forward to the PR, @wkyoshida! I’ll upload the datasets to iOS after and then start working on adding the references to them. We definitely need to have the SQLite feature finished to release this as we can’t add more data without it at this point, but having this done will make v2.2.0 all the closer :)

I think that referencing NLTK can definitely wait till we have something out that we can test, and then iterate from there 😊 Adding a popularity score into the output would be enough so that we can order it on the Swift side, as I figure ordering the results here would leave us open to it not being maintained when it’s added to the DB.

**Comment by andrewtavis** [comment]
@wkyoshida, I believe that [ef8b1e5](https://github.com/scribe-org/Scribe-Data/commit/ef8b1e5f35974f6673b90d184a405b6f55c74068) is the final touch of adding the output path to Scribe-iOS 😊 Will push the data to that repo now, and then we'll be set to start work on https://github.com/scribe-org/Scribe-iOS/issues/51 🚀

Let me know if you can think of anything else that's needed for the MVP for this 🙃 Really great work as always!

**Comment by wkyoshida** [comment]
> Adding a popularity score into the output would be enough so that we can order it on the Swift side, as I figure ordering the results here would leave us open to it not being maintained when it’s added to the DB.

Some thoughts around this are coming to mind now. I'll add them in #26 though

> Let me know if you can think of anything else that's needed for the MVP for this

I'm thinking of still adding some functionality here and there, e.g. #28 and nltk and etc, but I think we could be pretty good for the MVP! Or close (could depend on the additional thoughts I mentioned above)!
